### ISSUE TITLE
RuntimeError during mixed modality inference

### ISSUE THREAD

**Issue by in3der** [showlab/Show-o#90]
<img width="814" height="195" alt="Image" src="https://github.com/user-attachments/assets/e3dbfda7-e38d-4d7b-bc23-23ad52d77890" />

Hi, thank you for your amazing work on Show-o1 and Show-o2. The multimodal generation capabilities are truly impressive and inspiring.

I’m currently experimenting with your repository and used my own trained weights for the mixed modality task. The model was trained successfully, but I'm encountering a runtime error during inference using the inference_mixed_modality.py script.

I'm using the default prompts provided in your repository:
🔗 [show-o2/prompts/mixed_modal_prompts.txt](https://github.com/showlab/Show-o/blob/main/show-o2/prompts/mixed_modal_prompts.txt)

**Comment by Sierkinhane** [comment]
Hi, thank you for your interest in our work! It looks like no images are generated. Can you share your training loss?

**Comment by in3der** [comment]
Hello,
First of all, thank you very much for your kind and helpful response.

Based on your suggestions, we checked and confirmed that we had only completed the first stage of fine-tuning.
The training loss from this stage is as follows:

<img width="1304" height="635" alt="Image" src="https://github.com/user-attachments/assets/a3e3054a-8152-478d-bfe3-decbed58b97a" />

At this stage, the interleaved images were generated successfully, but the text outputs were not satisfactory.

We are now proceeding with the second stage (full fine-tuning).
However, we are encountering an issue where the loss becomes nan starting from step 2, and the training process crashes due to a timeout error around step 20.

This issue occurs not only with the weights fine-tuned from the first stage, but also with the official 1.5B pretrained weights you provided.
When transitioning to the second stage, we did not change any configuration or dataset settings except for unfreezing the frozen_params&lr. (The dataset used is AvaMERG.)

<img width="1325" height="717" alt="Image" src="https://github.com/user-attachments/assets/747918e6-a555-4236-8c87-ef79933d6d14" />

We would greatly appreciate any advice or insights you may have regarding this issue.
Thank you for taking the time to read this.

**Comment by Sierkinhane** [comment]
What's the loss value of step 1?

**Comment by Sierkinhane** [comment]
Have you modified some code? It appears that the learning rate will be adjusted; however, we have set a constant learning rate in the default setting.

**Comment by in3der** [comment]
```
lr_scheduler:
    scheduler:  # "constant_with_warmup" # "polynomial" 
    params:
        learning_rate: ${optimizer.params.learning_rate}
        warmup_steps: 0
        power: 0.5
```
I trained stage one by changing only the lr_scheduler to polynomial.
Then, in stage two, when using a constant schedule, I observed the following loss.

- the weights fine-tuned from the first stage

<img width="1125" height="131" alt="Image" src="https://github.com/user-attachments/assets/c11920ec-e0bc-49ea-9c3a-4a12a4b8f902" />

- 1.5B weight

<img width="1084" height="170" alt="Image" src="https://github.com/user-attachments/assets/92d7c504-3cfa-41eb-990f-64d7752fe3db" />

When applying a constant learning rate, the loss appeared as NaN, so I slightly adjusted the learning rate settings. However, a similar pattern appeared as shown in the above graph, so I am seeking advice.

**Comment by Sierkinhane** [comment]
It appears that no pre-trained weights were loaded. Please check if there are saved checkpoints and check the parameter `save_every` in the config file.

**Comment by in3der** [comment]
https://github.com/showlab/Show-o/blob/c4c863e2a19b7129458215c8c3990648c3dc2eff/show-o2/train_mixed_modality_simple.py#L164

After defining the model in the training code for stage two, we conducted a thorough comparison of all layer values between the fine-tuned weights and the 1.5B weights. We confirmed that all weights, except those frozen during stage one, were updated.

Please find attached the file containing the weight comparison.

[layer_differences.txt](https://github.com/user-attachments/files/21565804/layer_differences.txt)

**Comment by Sierkinhane** [comment]
You can see the loss values. The loss_flow converged to around 0.1 in stage 1, but the starting loss_flow was around 1, which means pre-trained weights were not loaded.

Please check if there are saved weights in the output folder after stage 1.

**Comment by in3der** [comment]
```
showo:
        model_name: "Showo2"
        load_from_showo: True
#        pretrained_model_path: "showlab/show-o2-1.5B"
        pretrained_model_path: "/home/a6000/bk-project/multimodal-showo/show-o2/1st_show-o2-1.5b-downstream-mixed-modality-432x432" 
        llm_model_path: "Qwen/Qwen2.5-1.5B-Instruct"
        llm_vocab_size: null # will be updated when setting the tokenizer in other code files
        hidden_size: 1536
        image_latent_dim: 16
        image_latent_height: 27
        image_latent_width: 27
        hq_image_latent_height: 64
        hq_image_latent_width: 64
        mixed_modal_latent_height: 27
        mixed_modal_latent_width: 27
        patch_size: 2
        num_diffusion_layers: 10
        clip_latent_dim: 1152
        add_qk_norm: True
        add_time_embeds: True
#        frozen_params: [ 'image_embedder_und', 'und_trans', 'showo', 'position_embedding'] 
        frozen_params: null
        params_not_load: null
```

<img width="1073" height="211" alt="Image" src="https://github.com/user-attachments/assets/fbe97aa3-c163-47c0-b0c8-ac0a8339944e" />

Thank you once again for your continued guidance.
I have adjusted the learning rate and modified access to the output folder.

As a result, the loss_flow started at 0.01.
However, starting from step 4, the loss becomes nan.

**Comment by Sierkinhane** [comment]
The loss values still look weird. It's not required to modify the pretrained_model_path. Can you check if there are checkpoints in this folder `1st_show-o2-1.5b-downstream-mixed-modality-432x432`?

**Comment by ahlee-smwu** [comment]
Firstly, thank you so much for your honest guidance.

I'm team worker of @in3der 's project, snd want to ask some more.

**1. We check up the checkpoint loading you mentioned below.**
Then we confirmed that the loss is started from the last checkpoint: (loss_ntp: 7.5625, loss_flow: 0.0159)
This is the result of our stage-1 training and saving path.
<img width="654" height="359" alt="Image" src="https://github.com/user-attachments/assets/99da4bcf-62d1-4240-97d3-4928482c123b" />

<img width="356" height="330" alt="Image" src="https://github.com/user-attachments/assets/97652a4f-e463-4ac1-b4aa-228da8b20201" />

This is the log for training stage-2 from our checkpoint.
: Loss is start from our checkpoint. So we confirm the weight loading has no matter.
<img width="1715" height="959" alt="Image" src="https://github.com/user-attachments/assets/e990378a-a60c-4e7c-b46a-907191f5739c" />

Also we are plan to do more training later as your recommend.

**2. Before developing stage-1, we need to confirm that the stage-2 training has no problem in our environment.**
As you explained, there are no reason for loss NAN for stage-2 training, even though we load our weak weight(that is not converged yet) from strage-1.

**3. For stage-2, we still get NAN loss.**
We loaded our weight(last checkpoint fo stage-1) then start stage-2 training from the loss of our weight.
But after few step, we get NAN for loss_flow
Both with show-o2-1.5B or our weight.

a) our weight
: same refult as before question of this issue.
<img width="1073" height="211" alt="Image" src="https://github.com/user-attachments/assets/04348f0b-3c6e-4c07-a514-ffe164ac1850" />

b) show-o2-1.5B weight
: same refult as before question of this issue.
<img width="1084" height="170" alt="Image" src="https://github.com/user-attachments/assets/00047b11-c611-48db-a5a3-b128278166b6" />

**4. We do some test for frozen weight.**
Only when parameter 'und_trans', 'showo' is opened, it occurs NAN of loss_flow.
So, in our opinion, may be attention mask or some other mask token have some problem.
As we look for the code, 'und_trans', 'showo' both use the attention mask.
So, can you check for the two parameters and attention mask?


Finally, we truly appreciate all your help. Thanks again!

**Comment by in3der** [comment]
Sure it's okay. Thank you for answering again

### 1. Dataset
- preprocessing AvaMERG dataset(empathy video) to interleaved dataset.
- Dataset length: 24696, each has around 4~6 pair of .png and txt
- num_update_steps_per_epoch: 6174 (for batch 1)
- huggingface link
https://huggingface.co/SIChoi/Show-o2_AvaMERG/tree/main

### 2. Configuration
**1) yaml file**

- `showo2_1.5b_downstream_mixed_modality_simple.yaml`
- The file is uploaded at huggingface.

**2) train file**

- we don't make matter change for train file, but we expect you would want to see this.
- So we uploaded this at huggingface too.

**3) terminal run command**

- `/home/ahlee/miniconda3/envs/showo2/bin/python /home/ahlee/miniconda3/envs/showo2/bin/accelerate launch --config_file ../accelerate_configs/8_gpus_deepspeed_zero2.yaml --main_process_port=9998 train_mixed_modality_simple.py config=configs/showo2_1.5b_downstream_mixed_modality_simple.yaml `

### 3. GPU
- We use NVIDIA RTX A6000(48GB) * 4
- Besides inference, we use this GPU setting for all training.
- Due to GPU memory limitation, we use 1 batch for interleaved dataset.

Additional note: During debugging, we confirmed that the NaN values first appeared in the last_hidden_states output right after passing through the `diffusion_head_a` layer.

We really appreciate all your help. Thank you

**Comment by ahlee-smwu** [comment]
Dear @Sierkinhane,

I’m reaching out to share an update regarding the NaN issue.
We switched the environment to a different GPU server, and this time obtained normal results without any NaNs.

All aspects, including the code, dataset, and configuration (batch size included), were identical to the previous server, as we fully cloned our Git repository. The only changes were:
- GPU: NVIDIA RTX A6000(48GB) * 4 -> NVIDIA RTX A100(80GB) * 4
- Conda environment is newly setted. Both envs is set with your given requirements.

We are currently investigating the cause of the NaNs.

We apologize that the issue was not related to your repository, but rather to our server environment.

Thank you again for all your help, and we will share our findings once we identify the cause.
